lib/rules/no-shadow.js

```diff
@@ -186,7 +260,7 @@ module.exports = {
 
                 if (shadowed &&
                     (shadowed.identifiers.length > 0 || (options.builtinGlobals && "writeable" in shadowed)) &&
-                    !isOnInitializer(variable, shadowed) &&
+                    !isOnInitializer(variable, shadowed) && !isInitPatternNode(variable, shadowed) &&
                     !(options.hoist !== "all" && isInTdz(variable, shadowed))
```

Comment:
This shouldn't change behavior when [`hoist`](https://eslint.org/docs/rules/no-shadow#hoist) option is set to `"all"`, since `hoist: "all"` means that shadowing should be reported regardless of where the variables are defined.

```js
/* eslint no-shadow: ["error", { hoist: "all" }] */

let x = foo((x, y) => {}); // Reports only y, but not x. Both should be reported.

let y;
```